btas/tensor.h

```diff
@@ -22,693 +22,564 @@ namespace btas {
-          using std::cend;
-          using std::begin;
-          using std::end;
-          range_ = range_type(x.range().lobound(), x.range().upbound());
-          if (&x.storage() != &this->storage()) { // safe to copy immediately, unless copying into self
-            array_adaptor<storage_type>::resize(storage_, range_.area());
-            std::copy(cbegin(x), cend(x), begin(storage_));
-          }
-          else {
-            // must use temporary if copying into self :(
-            storage_type new_storage;
-            array_adaptor<storage_type>::resize(new_storage, range_.area());
-            std::copy(cbegin(x), cend(x), begin(new_storage));
-            using std::swap;
-            swap(storage_,new_storage);
-          }
-          return *this;
-      }
+    /// Construct an evaluated tensor
+
+    /// This constructor will allocate memory for \c range.area() elements. Each element
+    /// will be initialized as:
+    /// \code
+    ///   for(auto&& idx: range)
+    ///     (*this)[idx] = op(*(it++));
+    /// \endcode
+    /// \tparam Range An input Range type.
+    /// \tparam InIter An input iterator type.
+    /// \tparam Op A unary operation type
+    /// \param range the input range type
+    /// \param first An input iterator for the argument
+    /// \param op The unary operation to be applied to the argument data
+    template <typename Range, typename InIter, typename Op>
+    Tensor(const Range& range, InIter it, const Op& op,
+           typename std::enable_if<btas::is_boxrange<Range>::value>::type* = 0)
+        : range_(range.lobound(), range.upbound()) {
+      auto size = range_.area();
+      array_adaptor<storage_type>::resize(storage_, size);
+      std::transform(it, it + size, begin(), op);
+    }
 
-      /// copy assignment
-      Tensor&
-      operator= (const Tensor& x)
-      {
-        range_ = x.range_;
-        storage_ = x.storage_;
+    /// copy constructor
+    /// It will accept Tensors and TensorViews
+    template <class _Tensor, class = typename std::enable_if<is_boxtensor<_Tensor>::value>::type>
+    Tensor(const _Tensor& x) : range_(x.range().lobound(), x.range().upbound()), storage_(x.cbegin(), x.cend()) {}
+
+    /// copy constructor
+    /// @note this makes a shallow copy of @п х if `storage_type` has shallow-copy semantics; if need a deep copy
+    ///       in that case use Tensor::clone()
+    /// @sa Tensor::clone()
+    Tensor(const Tensor& x) : range_(x.range()), storage_(x.storage_) {}
+
+    /// move constructor
+    Tensor(Tensor&& x) : range_(std::move(x.range())), storage_(std::move(x.storage_)) {}
+
+    /// @return deep copy of `*this`, even if `storage_type` is shallow copy
+    Tensor clone() const { return Tensor(range(), storage_type(storage().cbegin(), storage().cend())); }
+
+    /// returns element-wise conjugate of a *this tensor if have complex value type
+    std::conditional_t<is_complex_type_v<_T>, Tensor, const Tensor&> conj() {
```

Comment:
must be `const` function

Reply:
review your changes for const correctness

Reply:
added const in latest commit